src/ui/CategoryFacet/CategoryFacet.ts

```diff
@@ -0,0 +1,215 @@
+import { Initialization } from '../Base/Initialization';
+import { exportGlobally } from '../../GlobalExports';
+import { CategoryFacetTemplates } from './CategoryFacetTemplates';
+import { CategoryValueRoot } from './CategoryValueRoot';
+import { CategoryFacetQueryController } from '../../controllers/CategoryFacetQueryController';
+import { SVGDom } from '../../utils/SVGDom';
+import { SVGIcons } from '../../utils/SVGIcons';
+import { QueryStateModel } from '../../models/QueryStateModel';
+import 'styling/_CategoryFacet';
+import { IAttributesChangedEventArg, MODEL_EVENTS } from '../../models/Model';
+import { Utils } from '../../utils/Utils';
+import { CategoryValue } from './CategoryValue';
+import { contains, isArray } from 'underscore';
+import { Assert } from '../../misc/Assert';
+import { QueryEvents } from '../../events/QueryEvents';
+
+export interface CategoryFacetOptions {
+  field: IFieldOption;
+  title?: string;
+  id: string;
+}
+
+/**
+ * This component allows to render hierarchical facet values. It determines the filter to apply depending on the current path of values that
+ * are selected. The path is a a combination of values that follow each other in a hierarchy.
+ * TODO: Add examples. Explain what a path is.
+ */
+export class CategoryFacet extends Component {
+  public categoryFacetQueryController: CategoryFacetQueryController;
+  public listenToQueryStateChange = true;
+
+  static doExport = () => {
+    exportGlobally({
+      CategoryFacet
+    });
+  };
+
+  static ID = 'CategoryFacet';
+
+  static options: CategoryFacetOptions = {
+    field: ComponentOptions.buildFieldOption({ required: true }),
+    title: ComponentOptions.buildLocalizedStringOption({
+      defaultValue: l('NoTitle')
+    }),
+    id: ComponentOptions.buildStringOption({ postProcessing: (value, options: CategoryFacetOptions) => value || (options.field as string) })
+  };
+
+  private categoryValueRoot: CategoryValueRoot;
+  private categoryFacetTemplates: CategoryFacetTemplates;
+  private facetHeader: Dom;
+  private waitElement: Dom;
+  public queryStateAttribute: string;
+
+  public categoryValueRootModule = CategoryValueRoot;
+
+  constructor(public element: HTMLElement, public options: CategoryFacetOptions, bindings?: IComponentBindings) {
+    super(element, 'CategoryFacet', bindings);
+    this.options = ComponentOptions.initComponentOptions(element, CategoryFacet, options);
+
+    this.categoryFacetQueryController = new CategoryFacetQueryController(this);
+    this.categoryFacetTemplates = new CategoryFacetTemplates();
+    this.categoryValueRoot = new this.categoryValueRootModule($$(this.element), this.categoryFacetTemplates, this);
+
+    this.bind.onRootElement(QueryEvents.duringQuery, () => this.addFading());
+    this.bind.onRootElement(QueryEvents.deferredQuerySuccess, () => this.removeFading());
+    this.buildFacetHeader();
+    this.initQueryStateEvents();
+  }
+
+  /**
+   * Changes the active path and triggers a new query.
```

Comment:
I think it would be helpful to have some examples of what's a "path".

Reply:
Yes you're right. I plan on explaining in the documentation of the class what a path is and the different wording I use in the documentation of functions/attributes.